src/kdc/kdc_preauth.c

```diff
@@ -888,17 +885,24 @@ get_preauth_hint_list(krb5_kdc_req *request, krb5_kdcpreauth_rock rock,
     state->rock = rock;
     state->realm = rock->rstate->realm_data;
     state->e_data_out = e_data_out;
+    state->pa_data = NULL;
+    state->ap = preauth_systems;
 
-    state->pa_data = calloc(n_preauth_systems + 1, sizeof(krb5_pa_data *));
-    if (!state->pa_data) {
-        free(state);
-        (*respond)(arg);
-        return;
-    }
+    if (alloc_pa_data(KRB5_PADATA_FX_FAST, 0, &pa) != 0)
+        goto error;
+    if (add_pa_data_element(&state->pa_data, pa) != 0)
+        goto error;
+    if (add_etype_info(kdc_context, rock, &state->pa_data) != 0)
+        goto error;
```

Comment:
If `add_etype_info` fails, don't we end up freeing an already-added pa_data element?  These should probably happen in the other order.

Reply:
Only insofar as we free the whole state->pa_data list.  I'm not sure what order change is being proposed.